File: grind75/week1/043-01-Matrix-Medium/043-01-Matrix-Medium.py
# Question link: https://leetcode.com/problems/01-matrix/description/
# This question is a bit hard to understand at first, so here are some resources if I forget

# https://www.youtube.com/watch?v=HuZ3u2sRIPY
# Rotting orange problem is very similar in nature, probably easier to understand too
# https://github.com/gcmaidana/LeetCode-Solutions/blob/main/misc/994.%20Rotting%20Oranges%20(Medium)
# https://leetcode.com/problems/rotting-oranges/

class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:

        ROWS = len(mat)
        COLS = len(mat[0])
        q = deque()
        directions = [(0,1), (0,-1), (1,0), (-1,0)]

        for r in range(ROWS):
            for c in range(COLS):
                if mat[r][c] == 0:
                    q.append((r,c))
                else:
                    mat[r][c] = float('-INF')
        
        while q:
            row, col = q.popleft()
            for dr, dc in directions:
                r = dr + row
                c = dc + col
                if 0 <= r < ROWS and 0 <= c < COLS and mat[r][c] == float('-INF'):
                    q.append((r,c))
                    mat[r][c] = mat[row][col] + 1
        
        return mat
    


        

File: grind75/week1/043-01-Matrix-Medium/043-01-Matrix-Medium.cpp
// Problem link: https://leetcode.com/problems/01-matrix/description/

class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int ROWS = mat.size();
        int COLS = mat[0].size();
        std::queue<std::pair<int, int>> q;
        vector<vector<int>> directions = { {0,1}, {0, -1}, {1, 0}, {-1, 0} };


        for(int r = 0; r < ROWS; r++)
        {
            for(int c = 0; c < COLS; c++)
            {
                if(mat[r][c] == 0)
                {
                    q.push({r,c});
                }
                else
                {
                    mat[r][c] = -1;
                }
            }
        }

        while(!q.empty())
        {
            std::pair<int, int> front = q.front();
            int row = front.first;
            int col = front.second;
            q.pop();

            // check each direction and do something...
            for(const auto& direction : directions)
            {
                int dr = direction[0];
                int dc = direction[1];
                int r = row + dr;
                int c = col + dc;
                if(0 <= r && r < ROWS && 0 <= c && c < COLS && mat[r][c] == -1)
                {
                    q.push({r, c});
                    mat[r][c] = mat[row][col] + 1;
                }
            }
        }

        return mat;
        
    }
};
